File: tests/test_read_bumphunter_results.py
from __future__ import annotations

import os
import subprocess
from pathlib import Path

import pytest

_REPO_ROOT = Path(__file__).resolve().parents[1]
_BH_LOG_FIXTURE = _REPO_ROOT / "tests" / "root_macros" / "BHresults_sample.json"

# Thin Python/pytest wrapper invoking the ROOT test macro
# tests/root_macros/test_read_bumphunter_results.cpp via subprocess.run,
# matching the wrapper pattern already used for
# tests/test_plot_postfit_macro.py. `read_bumphunter_results()` needs a
# real ROOT/RooFit runtime this repository's own pytest dev venv does not
# have, so this sources scripts/setup_buildAndFit.sh itself, same as
# every other real-ROOT test in this repository.
_PROBE = r"""
repo_dir="$PWD"
source "$repo_dir/scripts/setup_buildAndFit.sh" >/dev/null
setup_status=$?

if (( setup_status != 0 )); then
    echo "setup_status=$setup_status"
    exit "$setup_status"
fi

root -l -b -q "tests/root_macros/test_read_bumphunter_results.cpp(\"$BH_LOG_FIXTURE\")"
"""


def _run_read_bumphunter_results_macro(fixture: Path) -> subprocess.CompletedProcess[str]:
    env = os.environ.copy()
    env["BH_LOG_FIXTURE"] = str(fixture)

    return subprocess.run(
        ["bash", "-lc", _PROBE],
        cwd=_REPO_ROOT,
        env=env,
        text=True,
        capture_output=True,
        check=False,
    )


@pytest.mark.requires_root
@pytest.mark.requires_analysis_dependencies
def test_read_bumphunter_results_matches_known_fixture_and_reports_missing_file() -> None:
    assert _BH_LOG_FIXTURE.exists(), "expected BHresults_sample.json fixture missing"

    completed = _run_read_bumphunter_results_macro(_BH_LOG_FIXTURE)

    assert completed.returncode == 0, (
        "test_read_bumphunter_results.cpp macro failed:\n"
        f"stdout:\n{completed.stdout}\n"
        f"stderr:\n{completed.stderr}"
    )
    assert "TEST_READ_BUMPHUNTER_RESULTS_OK" in completed.stdout, completed.stdout

File: tests/root_macros/test_read_bumphunter_results.cpp
// First-ever ROOT-macro unit test in this repository (see
// doc/TIER3_COMPLETION_PLAN.md Chunk 11 / doc/ACTIVITY_LOG.md's Tier 3
// Chunk 11.B entry): exercises read_bumphunter_results() in isolation,
// since it needs no TCanvas/graphics at all, against a small fixture
// BHresults.json (tests/root_macros/BHresults_sample.json - this
// repository's own J100 canonical run has none, since it is unmasked)
// plus a missing-file case. Invoked the same way plot_postfit.cpp's own
// macro-level test is: `root -l -b -q
// "tests/root_macros/test_read_bumphunter_results.cpp(\"<fixture>\")"`,
// from a thin Python/pytest wrapper
// (tests/test_read_bumphunter_results.py) matching the wrapper pattern
// already used for tests/test_plot_postfit_macro.py.

#include "../../plot_postfit.cpp"

void test_read_bumphunter_results(char const * bh_log_fixture) {

  bool ok{true};

  auto check_float = [&](char const * name, float actual, float expected) {
    if (fabs(actual - expected) > 1e-4f) {
      cout << "FAIL: " << name << " = " << actual << ", expected " << expected << endl;
      ok = false;
    }
  };

  BumpHunterInfo const info = read_bumphunter_results(bh_log_fixture);

  if (! info.available) {
    cout << "FAIL: expected available == true for existing fixture " << bh_log_fixture << endl;
    ok = false;
  }

  check_float("global_pval", info.global_pval, 0.1234f);
  check_float("significance", info.significance, 2.5f);
  check_float("mask_min", info.mask_min, 500.0f);
  check_float("mask_max", info.mask_max, 700.0f);

  string const missing_path = string(bh_log_fixture) + ".does_not_exist";
  BumpHunterInfo const missing_info = read_bumphunter_results(missing_path);

  if (missing_info.available) {
    cout << "FAIL: expected available == false for missing file " << missing_path << endl;
    ok = false;
  }

  check_float("missing global_pval", missing_info.global_pval, 0.0f);
  check_float("missing significance", missing_info.significance, 0.0f);
  check_float("missing mask_min", missing_info.mask_min, 0.0f);
  check_float("missing mask_max", missing_info.mask_max, 0.0f);

  if (! ok) {
    cout << "TEST_READ_BUMPHUNTER_RESULTS_FAILED" << endl;
    exit(1);
  }

  cout << "TEST_READ_BUMPHUNTER_RESULTS_OK" << endl;

}
